### PR TITLE
feat(api-key): Add api-key authentication to middleware

### PR DESCRIPTION
**Pull request by sradevski**
Also did a bit of a cleanup on the auth middleware. There should be no behavioral changes, just moved code around.

### CODE REVIEW

File: packages/medusa/src/utils/authenticate-middleware.ts [hunk 1]
```diff
@@ -32,50 +35,40 @@ export const authenticate = (
...
 
-    // @ts-ignore
-    const session: MedusaSession = req.session || {}
+    // We only allow authenticating using a secret API key on the admin
+    if (authTypes.includes(API_KEY_AUTH) && isAdminScope(authScope)) {
+      const apiKey = await getApiKeyInfo(req)
+      if (apiKey) {
+        ;(req as AuthenticatedMedusaRequest).auth = {
+          actor_id: apiKey.id,
+          // TODO: Should we set this to the creator of the token (apiKey.created_by), or do we leave it empty?
```
Comment (olivermrbl) [inline packages/medusa/src/utils/authenticate-middleware.ts]: I would just leave it empty for now

File: packages/medusa/src/utils/authenticate-middleware.ts [hunk 2]
```diff
@@ -32,50 +35,40 @@ export const authenticate = (
...
+        }
 
-    let authUser: AuthUserDTO | null = null
-    if (authTypes.includes(SESSION_AUTH)) {
-      if (
-        session.auth_user &&
-        stringEqualsOrRegexMatch(authScope, session.auth_user.scope)
-      ) {
-        authUser = session.auth_user
+        return next()
```
Comment (olivermrbl) [inline packages/medusa/src/utils/authenticate-middleware.ts:50]: question: Having `return next()` here means you cannot authenticate as a user on requests that carry a bearer token. Problably an edge case, but do we want to allow for that?

I would have thought just passing through all the configured auth types would make sense. Obviously, returning early if `req.auth` is set in one of them.

File: packages/medusa/src/utils/authenticate-middleware.ts [hunk 3]
```diff
@@ -104,6 +97,96 @@ export const authenticate = (
...
 }
 
+const getApiKeyInfo = async (req: MedusaRequest): Promise<ApiKeyDTO | null> => {
+  const authHeader = req.headers.authorization
+  if (!authHeader) {
+    return null
+  }
+
+  const [tokenType, token] = authHeader.split(" ")
+  if (tokenType.toLowerCase() !== "bearer" || !token) {
```
Comment (olivermrbl) [inline packages/medusa/src/utils/authenticate-middleware.ts]: thought(non-blocking): Regarding your comment on Basic vs Bearer vs Something Third, I wanted to link a [discussion](https://github.com/medusajs/medusa/discussions/4037) opened in the community some time ago. 

This discussion actually prompted us to add a specific header `x-medusa-access-token` for API key authentication, instead of the Bearer which commonly carries a JWT.

However, I like the idea of moving back to Bearer or Basic for API key authentication, because those are native authentication headers and are therefore by default handled appropriately by clients and proxies, e.g. avoiding any type of caching. 

Would be curios to hear what you think?

**Review by olivermrbl — APPROVED** [review]
LGTM, clean PR @sradevski. Left a couple of comments.

File: packages/medusa/src/utils/authenticate-middleware.ts (hunk 2 above)
Comment (sradevski) [reply packages/medusa/src/utils/authenticate-middleware.ts:50]: @olivermrbl This was intentional, but whether we want it like that or not I am happy to discuss 😄 

My thought-process was that we need to have some ordering/precendence when doing authentication. I thought an API key to be the most specialized case, and as such, should be the first on the list. If you haven't authenticated with an API key, then we fall back to other modes of authentication.

I think there is no right or wrong solution here, I personally just like not having too many options as it adds confusion to both the implementation side, as well as the user side. Let me know what you think though, I am not set on this approach :)

Comment (olivermrbl) [reply packages/medusa/src/utils/authenticate-middleware.ts:50]: > I thought an API key to be the most specialized case, and as such, should be the first on the list

Agree with the precedence and your priority. 

> then we fall back to other modes of authentication.

This is what I want to allow, but with the current implementation, I don't think this is possible, because we return `next()` regardless of the result of the API key auth no? In case API key authentication fails, we should proceed to try the other types of authentication.

One scenario I can think of where returning `next()` early could be an issue would be if you've configured your SDK in your client with an API key (and this client has regular user sign in too), and you choose to revoke that API key, then all your users would get 401s on attempted sign-in. Only when you have removed or updated the API key will your users be able to sign in again.

As I said, this is probably an edge case since you are unlikely to use the bearer token, if you also have regular user session sign-in, but thought I'd highlight it.

File: integration-tests/plugins/__tests__/api-key/admin/api-key.spec.ts [hunk 4]
```diff
@@ -39,6 +40,9 @@ describe("API Keys - Admin", () => {
 
   beforeEach(async () => {
     await createAdminUser(dbConnection, adminHeaders)
+
+    // Used for testing cross-module authentication checks
+    regionService.createDefaultCountriesAndCurrencies()
```
Comment (olivermrbl) [inline integration-tests/plugins/__tests__/api-key/admin/api-key.spec.ts]: question: Don't we need an `await` here?

File: packages/medusa/src/utils/authenticate-middleware.ts (hunk 3 above)
Comment (sradevski) [reply packages/medusa/src/utils/authenticate-middleware.ts]: @olivermrbl ok went through that discussion, I think some points are valid.

However, one thing that people miss about Bearer tokens is that they are not defined anywhere on what exactly they should carry. The do commonly carry JWT tokens as Bearer was added for Oauth 2.0, but they can literally carry anything that says "Give access to the bearer of this token". 

Now, I think it's still good to align with what people expect, rather than what is technically correct. I checked Stripe and they suggest Basic auth with the token, which I think is fine and we can do. This will just need another middleware to extract the basic credentials before calling the auth middleware, but it's not a big deal.

When it comes to session vs JWT authentication, it all depends on which tradeoffs you want to take. I think a pure cookie-based session auth is a bit outdated as it limits how this API can be called from eg. mobile apps (as that discussion points out). However, if you do want to allow revoking JWT tokens, for example, or you need a fine-grained authorization management, you'll end up with some session state regardless, so JWT token authentication gets a different meaning. That, however, is something we can discuss at a later point.

For now, I can just follow Stripe's approach and just use basic auth?

File: integration-tests/plugins/__tests__/api-key/admin/api-key.spec.ts (hunk 4 above)
Comment (sradevski) [reply integration-tests/plugins/__tests__/api-key/admin/api-key.spec.ts]: Good catch, we do!

File: packages/medusa/src/utils/authenticate-middleware.ts (hunk 1 above)
Comment (sradevski) [reply packages/medusa/src/utils/authenticate-middleware.ts]: Sounds good 👍

File: packages/medusa/src/utils/authenticate-middleware.ts (hunk 2 above)
Comment (sradevski) [reply packages/medusa/src/utils/authenticate-middleware.ts:50]: @olivermrbl If you see, we only call `return next()` only if the `if(apiKey)` is true, and that is only true if the API key is valid, so it will fallback to other modes of authentication, but I'll add it in the test scenario.

File: packages/medusa/src/utils/authenticate-middleware.ts (hunk 3 above)
Comment (olivermrbl) [reply packages/medusa/src/utils/authenticate-middleware.ts]: > For now, I can just follow Stripe's approach and just use basic auth?

This sounds good to me.

File: packages/medusa/src/utils/authenticate-middleware.ts (hunk 2 above)
Comment (olivermrbl) [reply packages/medusa/src/utils/authenticate-middleware.ts:50]: Oh man, I'm so sorry for going back and forth on this. Missed a curly bracket.

But I am glad that we agree 😄

Comment (sradevski) [reply packages/medusa/src/utils/authenticate-middleware.ts:50]: Added one more test where we try to use a revoked key

Comment (sradevski) [reply packages/medusa/src/utils/authenticate-middleware.ts:50]: @olivermrbl no worries, it could also have been a bug so thanks for pushing 😄

**Comment by sradevski** [comment]
@olivermrbl please check the last commit for the basic auth approach.

**Comment by olivermrbl** [comment]
LGTM
